design/fast_finality.md

```diff
@@ -0,0 +1,273 @@
+smart contracts that use `block.number` as a rudimentary clock (e.g. SushiSwap).
+
+The sequencer's host immediately distributes the light batch to all other nodes, who gossip it onwards to other nodes
+(ensuring the sequencer cannot restrict the distribution of light batches to specific nodes, provided one of the nodes 
+who received the light batch is honest). These light batches are not sent to be included on the L1.
+
+When a node receives a light batch, it checks that it has also stored the light batch's parent. If 
+not, it walks the chain backwards, requesting any light batches it is missing until it hits a stored light batch. In 
+the current design, it requests these light batches from random nodes; once a gossip protocol is implemented, it will 
+request the light batches from its known peers.
+
+The linkage of each light batch to its parent also ensures that the sequencer's host cannot feed the enclave a light 
+batch, use RPC requests to gain information about the contents of the corresponding transactions, then feed the enclave 
+a different light batch (e.g. one where the sequencer performs front-running) to be shared with peers. The enclave will 
+automatically reject a second light batch with the same parent.
+
+From the user's perspective, the transactions in the light batch are considered final (e.g. responses to RPC calls from
+the client behave as if the transactions were completely final).
+
+### Production of rollups
+
+Only the sequencer produces rollups. A rollup is produced whenever one of the following conditions is met:
+
+1. The number of transactions across all light-batches since the last rollup exceeds `x`
+2. The total size of all transactions across all light-batches since the last rollup exceeds `y`
+3. The number of blocks since the last rollup was produced exceeds `z`
+
+`x`, `y` and `z` are configurable per network. Rules (1) and (2) ensure that the rollup can fit within the Ethereum gas 
+limit. Rule (3) reduces the risk associated with the sequencer failing.
+
+To produce a rollup, the sequencer's host requests the creation of a rollup. The sequencer's enclave produces a rollup 
+containing all the light batches created since the last rollup, in a sparse Merkle tree structure. This rollup is 
+encrypted and signed, then sent to be included on the L1.
+
+### Discovery of rollups
+
+Nodes scan incoming L1 blocks for new rollups. They validate each new rollup by:
+
+* Checking that it is produced by the designated sequencer, based on the sequencer listed in the management contract
+* Checking that it contains all the light batches produced since the last rollup. Each light batch contains the number 
+  of the rollup that will contain it. Since the rollup is a sparse Merkle tree, proving non-inclusion of a given light 
+  batch is straightforward
+
+They then persist the rollup, so that they have a record of which light batches have been confirmed on the L1.
```

Comment:
we need a section to describe what happens when they discover something fishy

Reply:
Done